optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/solution/cloner/FieldAccessingSolutionCloner.java

```diff
@@ -60,10 +60,10 @@
 
     protected final SolutionDescriptor<Solution_> solutionDescriptor;
 
-    protected final ConcurrentMap<Class<?>, Constructor<?>> constructorCache = new ConcurrentHashMap<>();
-    protected final ConcurrentMap<Class<?>, List<Field>> fieldListCache = new ConcurrentHashMap<>();
-    protected final ConcurrentMap<Pair<Field, Class<?>>, Boolean> deepCloneDecisionFieldCache = new ConcurrentHashMap<>();
-    protected final ConcurrentMap<Class<?>, Boolean> deepCloneDecisionActualValueClassCache = new ConcurrentHashMap<>();
+    protected final Cache<Class<?>, Constructor<?>> constructorCache = new ConcurrentCache<>();
+    protected final Cache<Class<?>, List<Field>> fieldListCache = new ConcurrentCache<>();
+    protected final Cache<Pair<Field, Class<?>>, Boolean> deepCloneDecisionFieldCache = new ConcurrentCache<>();
+    protected final Cache<Class<?>, Boolean> deepCloneDecisionActualValueClassCache = new ConcurrentCache<>();
```

Comment:
A map to `Boolean` is effectively a `Set`. Shouldn't we have that instead?

Reply:
(Consider the performance impact of boxing and unboxing those `Booleans` into `boolean`s in the methods below.)

Reply:
Nevermind, realized my mistake. You actually need to cache the `false`s as they are a result of computation.

Reply:
It's a Boolean, not a boolean:
TRUE: calculation result is true
FALSE: calculation result is false
null: calculation hasn't been done yet